### ISSUE TITLE
Upgrade legacy packaging config to avoid deprecated, to-be-removed functionality, support modern standards/tooling and greatly simplify maintenance

### ISSUE THREAD

**Issue by CAM-Gerlach** [spyder-ide/spyder#17747]
For a long time now, I've been aware that the Spyder `setup.py` uses a bunch of deprecated, to-be-removed functionality, includes a substantial amount of dynamic complexity that is no longer needed, and doesn't support modern packaging standards. However, I'd put it off for a long time since I'd previously thought this would be a pretty big project and require great care, non-trivial workarounds (or difficult decisions on whether) to preserve the current behavior

However, after some of these issues coming up on #17408 and prompting a careful inspection of Spyder's current `setup.py` (as well as the experience successfully doing the same modernization for QtPy, Docrepr and my own packages), it actually looks to be pretty straightforward and have little to no negative impact on the actual output sdists/wheels, since essentially all the deprecated features that would be impacted already don't actually work/are used.

To note, #17408 sets the stage for this and is a prerequisite before I begin work here; in addition, via my suggested changes there, it implements proper testing of building and installing from wheels on various platforms and supports modern PEP 517 builds, to help ensure this doesn't introduce any regressions (of course, I'll be testing locally and carefully inspecting the built distribution packages for any deltas).

Mostly for my own reference, I've documented in detail the various items involved, divided into four phases: low hanging fruit cleanup, making dynamic metadata declarative, removing/replacing deprecated options, and supporting/testing up to date packaging infrastructure.

## Cleaning up low hanging fruit

There's a fair amount of low-hanging fruit cleanup that doesn't touch the functionality:

* [ ] Removing the old Python 2 __future__ import
* [ ] Removing the `io` import and just using the builtin `open` that's identical on Python 3
* [ ] `subprocess.call` -> `subprocess.run` since we're on Python >3.4, and `except` -> `except Exception` and maybe fixing the logic
* [ ] `WINDOWS_INSTALLER_NAME` also appears unused and can likely be removed
* [ ] Removing the "workaround to replace `os.path.relpath` on Python < 2.6 (thought maybe better to replace with `pathlib`, or just the existing `MANIFEST.in`, see below)
* [ ] We could replace all the hacky `osp` stuff with `pathlib`, but as most of that is redundant anyway and would be removed in the next step regardless, I doubt most of its worth it at this point as it adds effort and risk for minimal gain
* [ ] Some other smaller fixes/cleanup

## Porting dynamic to declarative metadata

As a bigger but more positively impactful relatively straightforward change but a small fraction of the hacky dynamic code can be replaced with much simpler, more robust, more performant and easier to maintain declarative `setup.cfg` config:

* [ ] Moving all the metadata that's currently just static to the `setup.cfg` (which is the vast majority of it); this can usually be automated by `setup-cfg-upgrade`, though for a `setup.py` this complicated it might need a bit of manual help on things like the long description (which can be all replaced with `long-description = file: README.md` in `setup.cfg`)
* [ ] Making `__version__` a string literal and using `version = attr: spyder.__version__` in `setup.cfg` (should be a much more minimal change than in QtPy, since Spyder doesn't use a separate `_version` file any more).
* [ ] Right now we have CI checks for the `MANIFEST.in`, but it isn't being used for actual wheels that users get (just the sdists) because `include_package_data` not `True` and the `setup.py` has a whole hacky function to iterate the whole file tree and enumerate the files to include. Instead, this can be listed decoratively in the `MANIFEST.in` using simple globs.
* [ ] Likewise, there's a whole manual `get_packages` function to populate `packages`, which has been deprecated for a long time now that `setuptools` has `find_packages` built-in automatically.
* [ ] The "sanity check" can be removed, because it is already handled in `python_requires` and means the minimum Python version is not single-sourced.

## Addressing deprecated items

This just leaves a few remaining deprecated items: `data_files`, `cmdclass` and `scripts` (aside from the dev constraints loosening); however, all three are deprecated and will be removed soon anyway, and none of them actually work as intended right now, so we need to do something about them regardless:

* [ ] `scripts` are replaced by entrypoints, but in fact they are actually `pop`ed from the `setup_args` and never actually passed to `setup()` at all, and are thus unused (and the `spyder` entrypoint replaces their function), so the code can simply be removed as it is never used (the files will still be in the sdist, for any downstreams consuming them)
* [ ] `cmdclass`, and the `distutils`'s `install_data` it uses, has been formally deprecated with `distutils` since Python 3.10 (and discouraged for at least half a decade longer) and will stop working in Python 3.12 because `distutils` will finally be removed. However, all its used for is just force-updating the desktop database after installation on some Linux systems (which makes sure Spyder shows up right away), only affects Linux users installing it through `pip` and not Anaconda or their distro, and doesn't even work anyway with `wheels` (which the `setup.py` itself notes, and `pip` always installed from by default), so its already essentially useless and can just be removed.
* [ ] Likewise, including `data_files` has been deprecated for a long time, and [has never worked properly with wheels](https://github.com/pypa/setuptools/discussions/2648), so it can also just be removed (like `scripts`, the files will still be in the sdist if downstreams/distros are doing anything with them)

## Packaging infrastructure updates

Also, there are some related updates that should be done, as we did for QtPy, to ensure the package works correctly with modern standards and is properly tested and working (which also helps validate that these changes, and any future ones, don't break anything):

* [ ] Add a `pyproject.toml` with PEP 518 metadata to support modern PEP 517 builds (which pip will require soon), follow up-to-date standards and switch from the deprecated, to be removed setuptools `__legacy__` builder to the more modern Setuptools `build_meta` builder.
* [ ] Ensure the tests fully test the built package, and work out any kinks (which #17408 already mostly resolves) based on the QtPy test config
* [ ] Update the `RELEASE.md` instructions to they are up to date with current best practices

## Other updates

There are a few other necessary packaging updates that can be done in separate followup PRs, as while they are both fairly straightforward, they don't touch the packaging config and rather just affect how packaging is handled at runtime; they both relate to migration from the deprecated `pkg_resources` to modern, supported replacements (that are already transitive dependencies) and allow us to no longer directly require `setuptools` at runtime:

* [ ] Replace the deprecated `pkg_resources` entrypoint loading calls with the modern stdlib/backported `importlib-metadata` (mostly) drop-in replacement
* [ ] Replace the deprecated `pkg_resources` version parsing with the modern standard `packaging` package

I can open separate issues for these once #17408 is merged, and can take care of them either before or after the `setup.py` migration.

**Comment by ccordoba12** [comment]
Thanks @CAM-Gerlach for taking care of this!

**Comment by CAM-Gerlach** [comment]
You always like to thank me before I've actually finished with things, jaja :joy:

Though to be fair, to quote GI Joe, "knowledge is half the battle", especially in this case where the actual changes should be quite straightforward now that I have a strong background in packaging, have gone through the process on other Spyder repos and I've investigated and detailed exactly what needs to be done.

**Comment by juliangilbey** [comment]
Hi!

There's also still a use of `distutils` in `setup.py`, which is now deprecated. 

Best wishes, Julian

**Comment by ccordoba12** [comment]
You're right, thanks for pointing that out.
